runtime/bcutil/cfreader.c

```diff
@@ -623,56 +642,102 @@ readAttributes(J9CfrClassFile * classfile, J9CfrAttribute *** pAttributes, U_32
 				 */
+				U_32 cursor = 0;
+
+				/* The minimum number of raw data bytes needed to create a bogus type_annotation is:
+				 * num_annotations (2 bytes)
+				 * target_type (1 byte)
+				 */
+				const U_32 minimumRawDataBytes = 3;
+
 				Trc_BCU_MalformedTypeAnnotation(address);
-				if (!ALLOC_ARRAY(annotations->rawAttributeData, length + 1, U_8)) {
+
+				/* use at least the minimum number of bytes, or add an extra slot for the illegal target_type */
+				annotations->rawDataLength = OMR_MAX(minimumRawDataBytes, length + 1);
+
+				/* special case for length = 1. To ensure index is incremented correctly, 
+				 * add this extra byte to the end of the raw data array. */
+				if (1 == length) {
+					annotations->rawDataLength++;
+				}
+
+				/* create raw data array. */
+				if (!ALLOC_ARRAY(annotations->rawAttributeData, annotations->rawDataLength, U_8)) {
 					return -2;
 				}
 				index = attributeStart; /* rewind to the start of the attribute */
-				NEXT_U16(annotations->rawAttributeData[0], index); /* put in the num_annotations */
-				NEXT_U8(annotations->rawAttributeData[2], index);
-				if (CFR_TARGET_TYPE_ErrorInAttribute != annotations->rawAttributeData[2]) {
-					/* insert an error marker */
-					annotations->rawAttributeData[3] = annotations->rawAttributeData[2];
-					annotations->rawAttributeData[2] = CFR_TARGET_TYPE_ErrorInAttribute;
-					annotations->rawDataLength = length + 1;
-				} else { /* the attribute is already marked bad */
-					NEXT_U8(annotations->rawAttributeData[3], index);
-					annotations->rawDataLength = length;
+
+				/* read num_annotations or set dummy if these bytes do not exist. */
+				if (length >= 2) {
+					NEXT_U16(annotations->rawAttributeData[0], index);
+				} else {
+					/* there should be at least one type_annotation here for the illegal 
+					 * entry that is being created. */
+					annotations->rawAttributeData[0] = 0;
+					annotations->rawAttributeData[1] = 1;
 				}
-				for (cursor = 4; cursor < annotations->rawDataLength; ++cursor) {
-					CHECK_EOF(1);
-					NEXT_U8(annotations->rawAttributeData[cursor], index);
+
+				/* set illegal target_type followed by raw data */
+				if (minimumRawDataBytes == annotations->rawDataLength) { /* final slot is reserved for illegal target_type */
+					annotations->rawAttributeData[2] = CFR_TARGET_TYPE_ErrorInAttribute;
+				} else {
+					NEXT_U8(annotations->rawAttributeData[2], index);
+
+					if (CFR_TARGET_TYPE_ErrorInAttribute != annotations->rawAttributeData[2]) {
+						/* insert an error marker */
+						annotations->rawAttributeData[3] = annotations->rawAttributeData[2];
+						annotations->rawAttributeData[2] = CFR_TARGET_TYPE_ErrorInAttribute;
+					} else { /* the attribute is already marked bad */
+						annotations->rawAttributeData[3] = 0;
+						if (index != end) {
+							NEXT_U8(annotations->rawAttributeData[3], index);
+						}
+						annotations->rawDataLength = length - 1; /* won't be using the extra slot */
```

Comment:
This is the case I'm most unsure of.  Maybe an example would help here? Or an assert that rawDataLength > 4, assuming that's the right condition